traffic_ops/testing/api/v4/cachegroups_parameters_test.go

```diff
@@ -155,3 +289,224 @@ func DeleteTestCacheGroupParameter(t *testing.T, cgp tc.CacheGroupParameterReque
+	}
+
+	opts.QueryParameters.Set("orderby", "id")
+	opts.QueryParameters.Set("limit", "1")
+	opts.QueryParameters.Set("offset", "1")
+	respWithOffset, _, err := TOSession.GetAllCacheGroupParameters(opts)
+	if err != nil {
+		t.Fatalf("cannot get cachegroup parameters with offset: %v - alerts: %+v", err, respWithOffset.Alerts)
+	}
+	cachegroupParametersWithOffset := respWithOffset.Response
+	if !reflect.DeepEqual(cachegroupParameters.CacheGroupParameters[1:2], cachegroupParametersWithOffset.CacheGroupParameters) {
+		t.Error("expected GET cachegroup parameters with limit = 1, offset = 1 to return second result")
+	}
+
+	opts.QueryParameters.Set("orderby", "id")
+	opts.QueryParameters.Set("limit", "1")
+	opts.QueryParameters.Set("page", "2")
+	respWithPage, _, err := TOSession.GetAllCacheGroupParameters(opts)
+	if err != nil {
+		t.Fatalf("cannot get cachegroup parameters with page: %v - alerts: %+v", err, respWithPage.Alerts)
+	}
+	cachegroupParametersWithPage := respWithPage.Response
+	if !reflect.DeepEqual(cachegroupParameters.CacheGroupParameters[1:2], cachegroupParametersWithPage.CacheGroupParameters) {
+		t.Error("expected GET cachegroup parameters with limit = 1, page = 2 to return second result")
+	}
+
+	opts.QueryParameters = url.Values{}
+	opts.QueryParameters.Set("limit", "-2")
+	resp, _, err = TOSession.GetAllCacheGroupParameters(opts)
+	if !alertsHaveError(resp.Alerts.Alerts, "limit parameter must be bigger than -1") {
+		t.Errorf("expected GET cachegroup parameters to return an error for limit is not bigger than -1, actual error: %v - alerts: %+v", err, resp.Alerts)
+	}
+
+	opts.QueryParameters.Set("limit", "1")
+	opts.QueryParameters.Set("offset", "0")
+	resp, _, err = TOSession.GetAllCacheGroupParameters(opts)
+	if !alertsHaveError(resp.Alerts.Alerts, "offset parameter must be a positive integer") {
+		t.Errorf("expected GET cachegroup parameters to return an error for offset is not a positive integer, actual error: %v - alerts: %+v", err, resp.Alerts)
+	}
+
+	opts.QueryParameters = url.Values{}
+	opts.QueryParameters.Set("limit", "1")
+	opts.QueryParameters.Set("page", "0")
+	resp, _, err = TOSession.GetAllCacheGroupParameters(opts)
+	if !alertsHaveError(resp.Alerts.Alerts, "page parameter must be a positive integer") {
+		t.Errorf("expected GET cachegroup parameters to return an error for page is not a positive integer, actual error: %v - alerts: %+v", err, resp.Alerts)
+	}
+}
+
+func SortTestCacheGroupParameters(t *testing.T) {
+	resp, _, err := TOSession.GetAllCacheGroupParameters(client.RequestOptions{})
+	if err != nil {
+		t.Fatalf("Expected no error, but got: %v - alerts: %+v", err, resp.Alerts)
+	}
+	cachegroupParameters := resp.Response
+	sortedList := make([]string, 0, len(cachegroupParameters.CacheGroupParameters))
+	for _, cgparameters := range cachegroupParameters.CacheGroupParameters {
+		if cgparameters.Parameter != nil {
+			sortedList = append(sortedList, strconv.Itoa(*cgparameters.Parameter))
```

Comment:
This line will segfault if Traffic Ops returned an improper response with null or undefined `parameter` property in any element.

Reply:
I hope this line will take care of the null check. 
if cgparameters.Parameter != nil{

Reply:
You're right, I missed that, somehow